dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/dataset/notifications/DataSetNotificationServiceTest.java

```diff
@@ -264,4 +295,50 @@ void testSendCompletionSMSNotification() {
             .getPhoneNumbers()
             .contains(PHONE_NUMBER));
   }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  void sendCompleteDataSetNotificationsTest() {
+    // setup
+    List<DataSetNotificationTemplate> emailTemplates = new ArrayList<>();
+    DataSetNotificationTemplate emailTemplateInternal = new DataSetNotificationTemplate();
+    emailTemplateInternal.setDataSetNotificationTrigger(
+        DataSetNotificationTrigger.DATA_SET_COMPLETION);
+    emailTemplateInternal.setDeliveryChannels(Sets.newHashSet(DeliveryChannel.EMAIL));
+    emailTemplateInternal.setNotificationRecipient(DataSetNotificationRecipient.USER_GROUP);
+    emailTemplateInternal.getDataSets().add(dataSetA);
+
+    emailTemplates.add(emailTemplateInternal);
+
+    CompleteDataSetRegistration registration = new CompleteDataSetRegistration();
+    registration.setSource(organisationUnitA);
+    when(dsntService.getCompleteNotifications(any())).thenReturn(emailTemplates);
+
+    I18nFormat format = Mockito.mock(I18nFormat.class);
+    when(i18nManager.getI18nFormat()).thenReturn(format);
+    when(format.formatPeriod(any())).thenReturn("2000-1-1");
+
+    when(renderer.render(
+            any(CompleteDataSetRegistration.class), any(DataSetNotificationTemplate.class)))
+        .thenReturn(notificationMessage);
+
+    when(configurationProvider.getServerBaseUrl()).thenReturn("https://dhis2.org");
+
+    // mock an email sender so its args can be inspected
+    Iterator<MessageSender> itr = Mockito.mock(Iterator.class);
+    when(messageSenders.iterator()).thenReturn(itr);
+    when(itr.hasNext()).thenReturn(true, false);
+    when(itr.next()).thenReturn(emailMessageSender);
+
+    ArgumentCaptor<String> footerCaptor = ArgumentCaptor.forClass(String.class);
+
+    // condition
+    subject.sendCompleteDataSetNotifications(registration);
+
+    // checks
+    verify(emailMessageSender)
+        .sendMessageAsync(any(), any(), footerCaptor.capture(), any(), any(), anyBoolean());
+    String footer = footerCaptor.getValue();
+    assertTrue(footer.contains("https://dhis2.org/dhis-web-messaging/#/SYSTEM/"));
```

Comment:
the message UID at the end of the URL is omitted from the assertion as its so deeply buried in the processing stack it would be too complicated to try to get injected (it's covered by another test though)